### ISSUE TITLE
[FR] Support training on a subset of a dataset's classes without editing annotation files

### ISSUE THREAD

**Issue by gboeer** [LibreYOLO/libreyolo#828]
## Description

A very common scenario I have to deal with often, is to train only on a subset of my annotation data, most notably only on some specific class labels. Correct me if I'm wrong, but as I see it there is currently no easy way to do this in libreyolo.

There are some workarounds but all of them are not optimal in a different way:

- leaving nc at the full count and just deleting the excluded class from `names` in the yaml looks like it should work but it doesn't: nothing  cross-references `names` against annotation class ids, so the  "excluded" class's boxes are still parsed and still trained on exactly as before
- setting nc to the smaller count while leaving annotation indices untouched crashes with a CUDA "scatter gather kernel index out of
  bounds" assertion instead of a clear error whenever a kept class's original index is >= the new nc (note that this is adressed by https://github.com/LibreYOLO/libreyolo/pull/827) 
- physically rewriting the annotation .txt files (drop/renumber lines), which is destructive to the source dataset and needs manual
  backup/versioning to ever recover the excluded class later.

## Use case

Given:
    names: {0: car, 1: bicycle, 2: dog, 3: cat}
    nc: 4

We want to train on {car, bicycle, cat} only, treating dog boxes as if they were never annotated (no loss contribution, no output slot for them), without modifying the dataset on disk.

**Comment by gboeer** [comment]
I've started working on this feature. If you're interested I can submit a PR in some time.

**Comment by EHxuban11** [comment]
Hello @gboeer 

Sorry, I forgot to answer. 
This is an excelent feature, and would be happy if you work on adding this. Other libraries also support this feature, so it's a gap of LibreYOLO: https://docs.ultralytics.com/modes/train/#train-settings

If you could maintain that same user facing API it would be great. 

<img width="476" height="80" alt="Image" src="https://github.com/user-attachments/assets/05a9e37c-c641-481b-841c-d7f0c0bd21b3" />

**Comment by gboeer** [comment]
Hi @EHxuban11  
yes indeed i followed the same api similar to this, ie supplying an additional classes attribute.

I am a bit unsure about one decision i made.
I decided to keep the output layer of the model the same size as is declared by the dataset.yaml.
This means, the classes will keep their id as has been defined in the yaml.
By using the classes attribute during train, only those class ids defined will be kept in the training set.
However, in theory this doesn't keep the model from predicting class ids that have not been trained (the output layer is still the size of the original classes).
Eventhough, this is unlikely after a good training run, technically it's possible.

The other option would be, to resize the output layer to the number of classes as defined in the classes attribute.
However, this will mean the original class ids, as have been defined by the dataset.yaml, are not preserved, ie they may have to be remapped to.
I did not go this way, because it basically invalidates the dataset definition, ie one can only get the class label->id mapping from the trained model itself.

Do you have an oppinion which route to take here?

**Comment by EHxuban11** [comment]
Hello @gboeer,

Keep the head at the dataset nc and only filter the labels, which is what you already did and how other libraries behave. The "untrained ids can still be predicted" concern is theoretical imo, but if a user wants a hard guarantee they can pass classes= at predict time to be sure.

You will also have to change the validator to exclude the ids we haven't trained on, otherwise they keep their ground truth and pull the mAP down.

So to make it short, your choice was right!

**Comment by EHxuban11** [comment]
Hello @gboeer , your contribution has been merged to dev!!

Thank you as always for your great work and for your patience through the weeks that this lasted.
